### DISCUSSION TITLE
What is the correct way to clean up the removal of a constraint?

### DISCUSSION THREAD

**Discussion by DennisSmolek** [jrouwe/JoltPhysics.js#159] [Q&A] [answered]
In my Jolt wrapper I try to manage all the cleanup of the various jolt objects. 
In my constraint system I first tried: 
`this.joltPhysicsSystem.RemoveConstraint(constraint);`

Which seemed ok but was causing some issues. 

So I tried:
`Jolt.destroy(constraint);`

This fixed the crashes I was having with `RemoveConstraint` but now when I try and cleanup the whole system,  I get crashes.

I can't quite confirm it yet, but what I think is happening is when the `RigidBodies` get their own `destroy()` call it happens before the constraint attempts to destroy or remove. And in their cleanup process they either also call destroy or need to access the bodies. I don't know..

For now I've commented out any removal/destroy process for constraints and this has stopped the crashes. 

Is there a better method to removing the constraint? Is it even necessary?

**Comment by LeXXik** [comment]
I only activate the bodies and do `.RemoveConstraint()`. No errors.

**Comment by jrouwe** [comment, marked as answer]
The proper sequence would be:

```
let body1 = <create and add a body>
let body2 = <create and add a body>

let constraintSettings = new Jolt.XXXConstraintSettings();
let constraint = constraintSettings.Create(body1, body2);
Jolt.destroy(constraintSettings);

physicsSystem.AddConstraint(constraint);

// simulate...

physicsSystem.RemoveConstraint(constraint);

// now you can remove and destroy body 1 and 2
```

**Reply by LeXXik** [reply]
Hmm, I would assume `Jolt.destroy(constraint);` would not be needed, since `constraint` is ref counted?

**Reply by jrouwe** [reply]
You're right, sorry. I've updated the code.
